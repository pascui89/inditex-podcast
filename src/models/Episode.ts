export interface Episode {
  episodeUrl: string;
  trackTimeMillis: number;
  artistIds: number[];
  collectionViewUrl: string;
  description: string;
  shortDescription: string;
  feedUrl: string;
  closedCaptioning: string;
  trackId: number;
  trackName: string;
  releaseDate: string;
  collectionId: number;
  collectionName: string;
  country: string;
  artworkUrl600: string;
  episodeFileExtension: string;
  artworkUrl160: string;
  episodeContentType: string;
  previewUrl: string;
  artworkUrl60: string;
  artistViewUrl: string;
  trackViewUrl: string;
  genres: { name: string; id: string }[];
  episodeGuid: string;
  kind: string;
  wrapperType: string;
}
