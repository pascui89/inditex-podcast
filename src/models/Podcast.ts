export interface Podcast {
  feed: {
    author: {
      name: {
        label: string;
      };
      uri: {
        label: string;
      };
    };
    entry: PodcastEntry[];
    updated: {
      label: string;
    };
    rights: {
      label: string;
    };
    title: {
      label: string;
    };
    icon: {
      label: string;
    };
    link: {
      attributes: {
        rel: string;
        type: string;
        href: string;
      };
    }[];
    id: {
      label: string;
    };
  };
}

export interface PodcastEntry {
  'im:name': {
    label: string;
  };
  'im:image': {
    label: string;
    attributes: {
      height: string;
    };
  }[];
  summary: {
    label: string;
  };
  'im:price': {
    label: string;
    attributes: {
      amount: string;
      currency: string;
    };
  };
  'im:contentType': {
    attributes: {
      term: string;
      label: string;
    };
  };
  rights: {
    label: string;
  };
  title: {
    label: string;
  };
  link: {
    attributes: {
      rel: string;
      type: string;
      href: string;
    };
  };
  id: {
    label: string;
    attributes: {
      'im:id': string;
    };
  };
  'im:artist': {
    label: string;
    attributes?: {
      href: string;
    };
  };
  category: {
    attributes: {
      'im:id': string;
      term: string;
      scheme: string;
      label: string;
    };
  };
  'im:releaseDate': {
    label: string;
    attributes: {
      label: string;
    };
  };
}

export const podcastMock: Podcast = {
  feed: {
    author: {
      name: {
        label: 'Author Name',
      },
      uri: {
        label: 'Author URI',
      },
    },
    entry: [
      {
        'im:name': {
          label: 'Podcast Name 1',
        },
        'im:image': [
          {
            label: 'Image URL 1',
            attributes: {
              height: '100',
            },
          },
          {
            label: 'Image URL 1',
            attributes: {
              height: '100',
            },
          },
          {
            label: 'Image URL 1',
            attributes: {
              height: '100',
            },
          },
        ],
        summary: {
          label: 'Podcast Summary 1',
        },
        'im:price': {
          label: '9.99',
          attributes: {
            amount: '9.99',
            currency: 'USD',
          },
        },
        'im:contentType': {
          attributes: {
            term: 'Term 1',
            label: 'Label 1',
          },
        },
        rights: {
          label: 'Podcast Rights 1',
        },
        title: {
          label: 'Podcast Title 1',
        },
        link: {
          attributes: {
            rel: 'Rel 1',
            type: 'Type 1',
            href: 'Link URL 1',
          },
        },
        id: {
          label: 'Podcast ID 1',
          attributes: {
            'im:id': '1',
          },
        },
        'im:artist': {
          label: 'Artist 1',
          attributes: {
            href: 'Artist URL 1',
          },
        },
        category: {
          attributes: {
            'im:id': '1',
            term: 'Category Term 1',
            scheme: 'Category Scheme 1',
            label: 'Category Label 1',
          },
        },
        'im:releaseDate': {
          label: 'Release Date 1',
          attributes: {
            label: 'Release Date Label 1',
          },
        },
      },
    ],
    updated: {
      label: 'Updated Label',
    },
    rights: {
      label: 'Feed Rights',
    },
    title: {
      label: 'Feed Title',
    },
    icon: {
      label: 'Feed Icon',
    },
    link: [
      {
        attributes: {
          rel: 'Rel 1',
          type: 'Type 1',
          href: 'Link URL 1',
        },
      },
    ],
    id: {
      label: 'Feed ID',
    },
  },
};
