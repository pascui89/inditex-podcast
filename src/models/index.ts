export type { Episode } from './Episode';
export type { Podcast, PodcastEntry } from './Podcast';
export type { PodcastDetail } from './PodcastDetail';
export type { ResponseData } from './ResponseData';
