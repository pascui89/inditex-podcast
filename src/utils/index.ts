export { millisecondsToMinutesSeconds, toDateFormat } from './dateUtils';
