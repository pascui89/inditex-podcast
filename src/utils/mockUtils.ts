import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const mockStore = configureStore([thunk]);
export const getMockStore = ( customStore = {} ) => {
    return mockStore({
        podcastReducer: {
            loading: true
        },
        episodeReducer: {
            loading: true,
            loadingEpisode: true
        },
        ...customStore
    });
}