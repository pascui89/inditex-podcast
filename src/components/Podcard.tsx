import { useId } from 'react';
import { useNavigate } from 'react-router-dom';

import { Avatar, Card, CardContent, Skeleton, Typography } from '@mui/material';
import { useTheme } from '@mui/material/styles';
import useMediaQuery from '@mui/material/useMediaQuery';

import { PodcastEntry } from '../models';

type Props = {
  item: PodcastEntry | null;
  loading?: boolean;
};

/**
 * Podcard Component.
 *
 * Displays a card representing a podcast item.
 *
 * @param {PodcastEntry} props.item - The podcast entry data.
 */
export default function Podcard(props: Props) {
  const id = useId();
  const { item } = props;
  const navigate = useNavigate();
  const theme = useTheme();
  const matches = useMediaQuery(theme.breakpoints.up('sm'));
  const { loading = false } = props;

  /**
   * Handle card click event.
   * Navigates to the podcast detail page.
   */
  const handleCardClick = () => {
    navigate(`/podcast/${props.item?.id.attributes['im:id']}`);
  };

  return (
    <Card
      key={id}
      sx={{
        cursor: 'pointer',
        transition: 'transform 0.2s',
        '&:hover': {
          transform: 'scale(1.05)',
          boxShadow:
            '0px 2px 1px -1px rgba(0,0,0,0.4), 0px 1px 1px 0px rgba(0,0,0,0.18), 0px 1px 3px 0px rgba(0,0,0,0.20)',
        },
      }}
      onClick={handleCardClick}
    >
      <CardContent
        sx={{
          maxWidth: 345,
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center',
          flexDirection: 'column',
        }}
      >
        {loading ? (
          <Skeleton
            animation="wave"
            variant="circular"
            width={matches ? 150 : 50}
            height={matches ? 150 : 50}
          />
        ) : (
          <Avatar
            sx={{ height: matches ? 150 : 50, width: matches ? 150 : 50 }}
            aria-label="recipe"
            src={(item && item['im:image'][2].label) || ''}
          />
        )}
        {loading ? (
          <Skeleton animation="wave" height={32} width="100%" />
        ) : (
          <Typography variant="h6" gutterBottom id="PodcastName">
            {item && item['im:name'].label}
          </Typography>
        )}
        {loading ? (
          <Skeleton animation="wave" height={28} width="100%" />
        ) : (
          <Typography variant="subtitle1" gutterBottom id="PodcastArtist">
            {item && item['im:artist'].label}
          </Typography>
        )}
      </CardContent>
    </Card>
  );
}
