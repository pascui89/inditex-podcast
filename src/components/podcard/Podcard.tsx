import { useId } from 'react';
import { useNavigate } from 'react-router-dom';

import { Avatar, Card, CardContent, Skeleton, Typography } from '@mui/material';
import { useTheme } from '@mui/material/styles';
import useMediaQuery from '@mui/material/useMediaQuery';

import { PodcastEntry } from '../../models';

type Props = {
  item: PodcastEntry | null;
  loading?: boolean;
};

/**
 * Podcard Component.
 *
 * Displays a card representing a podcast item.
 *
 * @param {PodcastEntry} item - The podcast entry data.
 */
export default function Podcard( { item, loading }: Props) {
  const id = useId();
  const navigate = useNavigate();
  const theme = useTheme();
  const matches = useMediaQuery(theme.breakpoints.up('sm'));

  /**
   * Handle card click event.
   * Navigates to the podcast detail page.
   */
  const handleCardClick = () => {
    navigate(`/podcast/${item?.id.attributes['im:id']}`);
  };

  return (
    <Card key={id} onClick={handleCardClick} className="cursor-pointer transition-transform duration-200 hover:scale-105 hover:shadow-lg">
      <CardContent className="max-w-[345px] flex justify-center items-center flex-col">
        {loading ? (
          <Skeleton
            data-testid="loading-data"
            animation="wave"
            variant="circular"
            width={matches ? 150 : 50}
            height={matches ? 150 : 50}
          />
        ) : (
          <Avatar
            sx={{ height: matches ? 150 : 50, width: matches ? 150 : 50 }}
            aria-label="recipe"
            src={(item && item['im:image'][2].label) || ''}
          />
        )}
        {loading ? (
          <Skeleton animation="wave" height={32} width="100%" />
        ) : (
          <Typography variant="h6" gutterBottom id="PodcastName" data-testid="podcastName">
            {item && item['im:name'].label}
          </Typography>
        )}
        {loading ? (
          <Skeleton animation="wave" height={28} width="100%" />
        ) : (
          <Typography variant="subtitle1" gutterBottom id="PodcastArtist">
            {item && item['im:artist'].label}
          </Typography>
        )}
      </CardContent>
    </Card>
  );
}
