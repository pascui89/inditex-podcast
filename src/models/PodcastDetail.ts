import { Episode } from './Episode';

export interface PodcastDetail {
  resultCount: number;
  results: Episode[];
}

export const mockPodcastDetail: PodcastDetail = {
  resultCount: 21,
  results: [
    {
      country: 'USA',
      artworkUrl600:
        'https://is3-ssl.mzstatic.com/image/thumb/Podcasts126/v4/31/d3/f2/31d3f28f-434c-45cc-17cf-ffcdfda0d78a/mza_9731489490781341652.jpg/600x600bb.jpg',
      feedUrl:
        'https://www.omnycontent.com/d/playlist/e73c998e-6e60-432f-8610-ae210140c5b1/9ff2dac3-12fd-4561-b512-ae33005f64f5/df6d181a-09ea-4bf2-adc6-ae33005f650d/podcast.rss',
      closedCaptioning: 'none',
      collectionId: 1096830182,
      collectionName: 'Drink Champs',
      artworkUrl60:
        'https://is3-ssl.mzstatic.com/image/thumb/Podcasts126/v4/31/d3/f2/31d3f28f-434c-45cc-17cf-ffcdfda0d78a/mza_9731489490781341652.jpg/60x60bb.jpg',
      artistViewUrl:
        'https://itunes.apple.com/us/artist/interval-presents/284341002?mt=2&uo=4',
      artworkUrl160:
        'https://is3-ssl.mzstatic.com/image/thumb/Podcasts126/v4/31/d3/f2/31d3f28f-434c-45cc-17cf-ffcdfda0d78a/mza_9731489490781341652.jpg/160x160bb.jpg',
      episodeContentType: 'audio',
      collectionViewUrl:
        'https://itunes.apple.com/us/podcast/drink-champs/id1096830182?mt=2&uo=4',
      trackTimeMillis: 9813000,
      episodeFileExtension: 'mp3',
      genres: [{ name: 'Music', id: '1310' }],
      episodeGuid: 'e7668686-4c6c-4176-90ba-b01b0112eaf5',
      description: 'N.O.R.E. & DJ EFN are the Drink Champs...',
      shortDescription: '',
      releaseDate: '2023-06-09T07:01:00Z',
      previewUrl:
        'https://chrt.fm/track/BE7515/traffic.omny.fm/d/clips/e73c998e-6e60-432f-8610-ae210140c5b1/9ff2dac3-12fd-4561-b512-ae33005f64f5/e7668686-4c6c-4176-90ba-b01b0112eaf5/audio.mp3?utm_source=Podcast&in_playlist=df6d181a-09ea-4bf2-adc6-ae33005f650d',
      artistIds: [284341002],
      trackViewUrl:
        'https://podcasts.apple.com/us/podcast/episode-367-w-de-la-soul/id1096830182?i=1000616244153&uo=4',
      episodeUrl:
        'https://chrt.fm/track/BE7515/traffic.omny.fm/d/clips/e73c998e-6e60-432f-8610-ae210140c5b1/9ff2dac3-12fd-4561-b512-ae33005f64f5/e7668686-4c6c-4176-90ba-b01b0112eaf5/audio.mp3?utm_source=Podcast&in_playlist=df6d181a-09ea-4bf2-adc6-ae33005f650d',
      trackId: 1000616244153,
      trackName: 'Episode 367 w/ De La Soul',
      kind: 'podcast-episode',
      wrapperType: 'podcastEpisode',
    },
  ],
};
