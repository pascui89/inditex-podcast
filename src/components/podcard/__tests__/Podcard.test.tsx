import { render, screen } from '@testing-library/react';

import Podcard from '../Podcard';
import { podcastMock } from '../../../models/Podcast';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import React from 'react';

global.React = React;

import { getMockStore } from '../../../utils/mockUtils';

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}));

describe('Podcard E2E Test', () => {  
  it('should renders episode details', () => {
    const { container } = render(
      <Provider store={getMockStore()}>
        <BrowserRouter>
          <Podcard item={podcastMock.feed.entry[0]} />
        </BrowserRouter>
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
  it('should display the correct podcast name and artist', () => {
    render(
      <Provider store={getMockStore()}>
        <BrowserRouter>
          <Podcard item={podcastMock.feed.entry[0]} />
        </BrowserRouter>
      </Provider>
    );
    const element = screen.getAllByRole('heading');
    expect(element).not.toHaveLength(0);
  });
});
