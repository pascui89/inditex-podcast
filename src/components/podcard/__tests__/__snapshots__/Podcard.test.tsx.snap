// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Podcard E2E Test should renders episode details 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root cursor-pointer transition-transform duration-200 hover:scale-105 hover:shadow-lg css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root max-w-[345px] flex justify-center items-center flex-col css-46bh2p-MuiCardContent-root"
    >
      <div
        aria-label="recipe"
        class="MuiAvatar-root MuiAvatar-circular css-1fsl38a-MuiAvatar-root"
      >
        <img
          class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
          src="Image URL 1"
        />
      </div>
      <h6
        class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
        data-testid="podcastName"
        id="PodcastName"
      >
        Podcast Name 1
      </h6>
      <h6
        class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom css-1wniyei-MuiTypography-root"
        id="PodcastArtist"
      >
        Artist 1
      </h6>
    </div>
  </div>
</div>
`;
